### PR TITLE
Project restructuring and encrypted payloads

### PR DESCRIPTION
**Pull request by achilleas-k**
*Counterpart to G-Node/gogs#47*

#### Project restructuring

Most of the restructuring parts were already pushed to the dev branch months ago.  I squashed a few WIP commits and I'm including them here for merging into master.  This refers to all commits up to and including 6bf3b95 (No mixed indent JSON strings).

Changes:
- go mod enabled.
- Added Dockerfile.
- Configuration is now performed through environment variables.
    - For containerised deployment, this makes it easy to specify an env file and load it into the container.
- All source code moved to  `cmd/gindex` subdirectory (see 2f39eb2 for reasoning).
- Vendored dependencies.
- Variable renaming (typos, unconventional names).
- Single configuration struct is set in main() during startup and is passed to all route handlers.
- Shared functionality moved to G-Node/libgin project (more to come).
- Functions no longer exported.
    - This project should serve only as the gin-dex service.  Any functionality that might be shared with other GIN projects should be in libgin.  There are no subpackages here that can be imported by other projects.
- Indexing requests are handled asynchronously.  A number of workers (currently hardcoded to 20) are started on service launch, indexing requests are put on a queue (buffer length hardcoded to 1000) and are run in order.  The calling side gets an OK (200) response as soon as the payload is decoded and the request is put on the queue.  No part of response indicates whether the indexing was successful.

#### Encrypted payloads

Changes in GOGS made in the interest of security do not allow API calls using cookies and CSRF tokens (see gogs/gogs@3db9b06).  This PR changes the way search and indexing requests work in the way described in Issue #2.

Changes:
- Key configuration option is a shared key between gin-dex and gogs used to symmetrically encrypt request payloads.
- gin-dex decrypts request payloads before unmarshalling and encrypts responses after marshalling.

### CODE REVIEW

File: cmd/gindex/indexObjects.go [hunk 1]
```diff
@@ -53,40 +54,43 @@ func (c *IndexCommit) ToJson() ([]byte, error) {
...
 	indexid := GetIndexCommitId(id.String(), bl.GinRepoId)
-	f_type, blobBuffer, err := BlobFileType(bl)
+	ftype, blobBuffer, err := BlobFileType(bl)
 	if err != nil {
 		log.Errorf("Could not determine file type: %v", err)
 		return nil
 	}
-	switch f_type {
+
+	repopath := cfg.RepositoryStore
```
Comment (mpsonntag) [inline cmd/gindex/indexObjects.go]: Rename to a more unambiguous variable name.

File: cmd/gindex/indexObjects.go [hunk 2]
```diff
@@ -147,7 +151,7 @@ func (bl *IndexBlob) AddToIndex(server *ESServer, repopath string, id gig.SHA1)
 	if err != nil {
 		return err
 	}
-	err = AddToIndex(data, server, server.blindex, "blob", indexid)
+	err = AddToIndex(data, cfg.Elasticsearch, cfg.Elasticsearch.blindex, "blob", indexid)
```
Comment (mpsonntag) [inline cmd/gindex/indexObjects.go:152]: Redundant signature variables.

File: cmd/gindex/handler.go [hunk 3]
```diff
@@ -123,28 +125,29 @@ func SuggestH(w http.ResponseWriter, r *http.Request, els *ESServer, gins *GinSe
...
 	rbd := IndexRequest{}
 	err := getParsedBody(r, &rbd)
 	log.Debugf("Got an indexing request: %+v", rbd)
 	if err != nil {
 		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
-	err = ReIndexRepoWithPath(fmt.Sprintf("%s/%s", *rpath, strings.ToLower(rbd.RepoPath)+".git"),
-		"master", els, rbd.RepoID, rbd.RepoPath)
+	err = ReIndexRepoWithPath(cfg, fmt.Sprintf("%s/%s", cfg.RepositoryStore, strings.ToLower(rbd.RepoPath)+".git"), "master", rbd.RepoID, rbd.RepoPath)
```
Comment (mpsonntag) [inline cmd/gindex/handler.go]: Redundant signature variables.

File: cmd/gindex/main.go [hunk 4]
```diff
@@ -11,6 +11,81 @@ import (
...
+	// timeout for adding contents to index (in seconds)
+	timeout, err := strconv.ParseInt(libgin.ReadConfDefault("timeout", "60"), 10, 64)
+	if err != nil {
+		log.Printf("Error while parsing timeout variable: %v", err)
+		timeout = 60
+		log.Printf("Using default: %d", timeout)
+	}
+	cfg.Timeout = timeout
+
+	portstr := libgin.ReadConfDefault("port", "8099")
```
Comment (mpsonntag) [inline cmd/gindex/main.go]: The current dockerfile states 10443.

File: cmd/gindex/util.go [hunk 5]
```diff
@@ -16,22 +16,30 @@ import (
...
+	reqdata, err := libgin.DecryptString(key, string(encdata))
 	if err != nil {
-		log.Debugf("Could not unmarshal request [%s]: %v", string(data), err)
+		log.Errorf("Failed to decrypt request: %v", err)
+		return err
+	}
+
+	err = json.Unmarshal([]byte(reqdata), obj)
+	if err != nil {
+		log.Debugf("Could not unmarshal request [%s]: %v", string(reqdata), err)
```
Comment (mpsonntag) [inline cmd/gindex/util.go]: Should probably be `log.Errorf`.

File: cmd/gindex/handler_test.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
-func TestIndexHandler(t *testing.T) {
+func TestindexHandlerandler(t *testing.T) {
```
Comment (mpsonntag) [inline cmd/gindex/handler_test.go]: Search and replace gone wild.

File: cmd/gindex/util.go [hunk 7]
```diff
@@ -68,6 +68,20 @@ func getParsedHttpCall(method, path string, body io.Reader, token, csrfT string,
...
+func encodeResponse(obj interface{}, key []byte) ([]byte, error) {
+	data, err := json.Marshal(obj)
+	if err != nil {
+		log.Errorf("Failed to marshal response: %v", err)
+		return nil, err
+	}
+
+	encdata, err := libgin.EncryptString(key, string(data))
+	if err != nil {
+		log.Errorf("Failed to encrypt response: %v", err)
```
Comment (mpsonntag) [inline cmd/gindex/util.go:80]: Probably should `return nil, err`.

File: cmd/gindex/handler.go [hunk 8]
```diff
@@ -10,33 +10,89 @@ import (
...
-	repo := strings.ToLower(rbd.RepoPath)
+	data := idxTask.data
+	repo := strings.ToLower(data.RepoPath)
 	if repo[len(repo)-4:] != ".git" {
 		repo = repo + ".git"
 	}
-	repoidstr := fmt.Sprintf("%d", rbd.RepoID) // TODO: Make repoid int64 everywhere
-	err = IndexRepoWithPath(cfg, fmt.Sprintf("%s/%s", rpath, repo), "master", repoidstr, rbd.RepoPath)
+	repoidstr := fmt.Sprintf("%d", data.RepoID) // TODO: Make repoid int64 everywhere
+	err := IndexRepoWithPath(cfg, fmt.Sprintf("%s/%s", rpath, repo), "master", repoidstr, data.RepoPath)
```
Comment (mpsonntag) [inline cmd/gindex/handler.go:36]: Redundant signature variables.

File: cmd/gindex/handler.go [hunk 9]
```diff
@@ -10,33 +10,89 @@ import (
...
+		}()
+	}
+	log.Debug("%d workers ready and waiting\n", idxQueue.nWorkers)
+}
+
+// AddTask creates an IndexTask and adds it to the worker queue.
+func (idxQueue *IndexQueue) AddTask(req *http.Request, cfg *Configuration) error {
+	// Read and decrypt request data
+	data := libgin.IndexRequest{}
+	err := getParsedBody(req, cfg.Key, &data) // error
```
Comment (mpsonntag) [inline cmd/gindex/handler.go]: Remove comment.

File: cmd/gindex/handler.go [hunk 10]
```diff
@@ -10,33 +10,89 @@ import (
...
+	idxQueue.jobs <- &idxTask
+	return nil
+}
+
+// Handler for indexing requests.  Adds new indexing requests to the IndexQueue and logs any errors.
+func indexHandler(w http.ResponseWriter, r *http.Request, cfg *Configuration, idxQueue *IndexQueue) {
+	err := idxQueue.AddTask(r, cfg)
+	if err != nil {
+		log.Errorf("Error while preparing indexing task: %v", err)
 		w.WriteHeader(http.StatusInternalServerError)
```
Comment (mpsonntag) [inline cmd/gindex/handler.go:93]: Add `return` to not overwrite the Error header.

**Review by mpsonntag — APPROVED** [review]
Very nice!

File: cmd/gindex/indexObjects.go (hunk 2 above)
Comment (achilleas-k) [reply cmd/gindex/indexObjects.go:152]: Not redundant.  The third arguments (`blindex`) is either the Blob index or the Commit index and there could be more if we start indexing other types of data, so this is not always the same value.

**Pull request merged**